### ISSUE TITLE
Any suggestion to reproduce your reseults?

### ISSUE THREAD

**Issue by SiyuanHuang95** [vimalabs/VIMA#8]
Hi,

Thanks for your sharing! 

It seems that only one inference demo script is released and not much information w.r.t the training procedure. It would be great that you could offer us some insights about the training procedure. Like hardware, (a100 * 8) and time cost?

Bests,

**Comment by yunfanjiang** [comment]
Thank you for your interest in our work! Because the training code is heavily coupled with internal codebase, we don't have any further release plan. We will let you know if situations change.

That being said, we can still share that we build our training infrastructure based on Pytorch-Lightning with DDP. All experiments were conducted on cluster nodes, each with 8 V100 GPUs. Training the largest model took approximately one day. I hope this information can help. Let me know if you have further questions!

**Comment by SiyuanHuang95** [comment]
Hi, thanks for your reply. That is indeed helpful!

I still have some minor questions:

1. Mask RCNN. I noticed that you mentioned that the segmentation model needs to be fine-tuned. How many data samples you used for training that segmentation model, also full size? And the training step is separate from the policy training, and will it be frozen when training the policy? 

2. Is the "image of the single object" in the prompt prior knowledge? In other words, is it given by ground truth (like from the tenv simulator)?

3. The trainable parameters are policy? And the t5-related and the segmentation model should be frozen?

4. I am unclear about the training loss. The paper said it should be summed NLL loss for predicted actions. And the action space is like "pick and place" space, e.g. SE(2) for End-effector. My question is that, for a sequence like 4 frames, the training sample only has one pair of "P&P", how to calculate the loss then? Given the current observation, the model should always predict a consistent P&P across 4 frames. Did I misunderstand something?

5. Following 4th, saying the robot did not give ideal action, the offline training dataset should diverge with the actual ones from  env+action?

**Comment by yunfanjiang** [comment]
Thanks for the questions.

1. We fine-tuned it on a subset of agent training dataset. It contains 100 trajectories for each task, resulting in 22741 images and 61822 annotations in total. Yes, the fine-tuning stage was conducted separately. It was then frozen during policy training.
2. Prompt images' types are known once a task is given. In practice, we still obtain tokens in the same way as full scene images, except that bounding box coordinates are somewhat meaningless in this case.
3. The model size reported in scalability experiment only includes the sequence modeling part (i.e., those transformer layers). If we don't count the separate detection model which only needs to be trained once, all trainable components include the ViT image encoder, bounding box coordinate MLP, the adapter MLP between visual tokens and T5, last two layers of T5, transformer policy backbone, action decoders, as well as learnable positional embeddings.
4. In your example, given 4 steps of observations, we assume 4 action labels as well. The loss is averaged over 4 steps then.
5. I didn't fully understand this question. I assume you were asking about the compounding error, which is a common problem of imitation learning from expert demonstrations. In our case, we observed our model sometimes can correct its previous mistakes. Hope this is helpful.

**Comment by SiyuanHuang95** [comment]
Great, thanks for your reply!

1. Got it. And it seems the released version is without a perception module. And the object information is directly available from the simulator in the released version.
2. The "Prompt Image" from the task description is known when the task is given, the representation is coming from the captured images (front or bev), or coming from ground truth, like cad-style?
3. Clear! Thanks!
4. I am still confused for this part. From the dataset you provide, the action label is <pose_0_pos+rot> and <pose_1_pos+rot>. However, the corresponding observations could be more than 2, like 5 in some rearrange_then_restore tasks. In such a situation, how to compute loss is not quite clear.
5. Cool. LfD or IL is useful but sometimes not robust to that error. Quite interesting to see the performance from VIMA.

Your reply is quite helpful! Thanks again.

BTW, would it be possible to have you IM or slack? Or just leave it in GitHub. :-)

**Comment by yunfanjiang** [comment]
Hi @SiyuanHuang95, regarding the followup:

1. The current script directly queries ground-truth segmentation from the simulator. This is the oracle version in our paper. The codebase for the detection model needs further cleanup and we may release in the future. That being said, it's relatively easy to fine-tune your own from Detectron2.
2. All prompt images regardless of their types come from captured images.
3. Great.
4. Observations and actions should have the same length. Could you point me to the wrong sample you mentioned?
5. Sounds good.

Just leave messages on GitHub is fine. This discussion may help others as well.

**Comment by SiyuanHuang95** [comment]
Hi @yunfanjiang, thanks for your quick reply.

1~3 & 5: Clear. Thanks!
4: Okay. The dataset structure is reasonable when the first dim is the timestamp. When there are three observations (images), should there be 2 actions since the last image is the final result and no further action need to be taken? And my question is that, how let the policy be aware of the "steps" it needs to take? For training, the step only averages the total loss, e.g., divided by the pre-defined step number. For inferencing, the step is not used since the simulator decides whether the action is done. The training procedure is converted from sequence-level to image-level in such a case. Do I understand that correctly? If not, can you give some suggestions to train on the varied length (step) clips?
6. I noticed that you use the following line to scale the rotation; any idea for that?
```
actions["pose0_rotation"] = actions["pose0_rotation"] * 2 - 1
```
7. I Hope this issue will be helpful for anybody interested in robot learning related. :-)

Bests,

**Comment by SiyuanHuang95** [comment]
EDIT for training:

I was wrong in my last comment about image-level training. The policy also needs the history tokens, so we must model it with the sequence level.

**Comment by yunfanjiang** [comment]
Thanks for the followup questions @SiyuanHuang95 

4. > When there are three observations (images), should there be 2 actions since the last image is the final result and no further action need to be taken? 

In our dataset, we do not include terminal observations and hence obs and action sequences have the same length.

4. > how let the policy be aware of the "steps" it needs to take or does it need to be aware?

The agent doesn't need to know this information. All tasks have maximum number of steps. A task terminates once exceeding this limit.

6. Rotations are represented with quaternion in our case. Each number in a quaternion falls in the range of [-1, 1]. Since the model outputs rotations in [0, 1]. We scale and shift them to the target range.

7.  Yes but no. Models are only trained with data corresponding to the first eval level (placement generalization).

**Comment by yunfanjiang** [comment]
Hi @SiyuanHuang95, regarding your new questions.

4. Yes, you are right. We discard the terminal observation during training.
5. Training data was generated on the L1 level to evaluate new placement generalization. So the pickled file doesn't need to include info about the eval level. In terms of data splitting, we used 600K trajectories for training and the remaining 50K for validation purpose.
6. Held-out 4 tasks are reserved for the L4 evaluation.
7. The provided script demonstrates the evaluation process. Ours is a vectorized implementation for better throughput.

**Comment by SiyuanHuang95** [comment]
Hi Yunfan,

Thanks for your reply!

**Comment by dingjiansw101** [comment]
> Thank you for your interest in our work! Because the training code is heavily coupled with internal codebase, we don't have any further release plan. We will let you know if situations change.
> 
> That being said, we can still share that we build our training infrastructure based on Pytorch-Lightning with DDP. All experiments were conducted on cluster nodes, each with 8 V100 GPUs. Training the largest model took approximately one day. I hope this information can help. Let me know if you have further questions!

Hi Yunfan, how many nodes did you use for the training?
